File: web/src/lib/pools/names.ts
/**
 * The name a new pool arrives with, and the dice that roll another one.
 *
 * A blank name field is the first thing an operator meets in the wizard, and
 * the name they invent to get past it -- "test", "pool1" -- is then in every
 * runner name, every audit line and every `runs-on` for the life of the fleet.
 * Renaming a pool later does not un-write the workflows that already point at
 * it. So the wizard fills one in instead.
 *
 * The name it fills in is the pool's shape -- `zoomies-4vcpu-ubuntu-2404`, the
 * grammar in docs/naming.md -- because that is what a workflow author reading a
 * `runs-on` needs to know before they send a job to it, and because it is the
 * same name `zoomies init` suggests for a machine of that size. The wizard used
 * to lead with a kennel word instead, which left an operator who met both with
 * two conventions for one thing.
 *
 * The spaniel is still here, and still earns its place: it is what tells two
 * pools of the same shape apart, and what names a pool created before anything
 * about it is known. The brand prefix is not optional either -- it is what marks
 * a runner in the GitHub UI as ours rather than GitHub's.
 */

import type { BackendKind, Host } from '$lib/api/types';
import { RUNNER_NAME_PREFIX, sanitizeLabel } from '$lib/brand';

/**
 * Cocker spaniels, as cocker spaniels are actually named.
 *
 * The mark is a cocker spaniel doing zoomies, so the fleet is a kennel. Every
 * word here is one segment of lowercase letters -- no spaces, nothing to
 * sanitise away -- and short enough that the infrastructure half of the name
 * survives the length budget below.
 */
export const KENNEL: readonly string[] = [
  'banjo',
  'biscuit',
  'boogie',
  'bramble',
  'bubbles',
  'cocoa',
  'crumpet',
  'custard',
  'digby',
  'disco',
  'flapjack',
  'gizmo',
  'hazel',
  'jellybean',
  'jitterbug',
  'maple',
  'marmalade',
  'muffin',
  'noodle',
  'pancake',
  'pepper',
  'pickles',
  'popcorn',
  'rascal',
  'rocket',
  'rusty',
  'scampi',
  'toffee',
  'truffle',
  'waffles',
  'wiggles',
  'ziggy',
];

/**
 * How long a generated name may be.
 *
 * `sanitizeLabel` truncates at 40 characters, so a longer name would produce a
 * label that is a chopped-off version of it -- two strings an operator has to
 * hold in their head at once, for no gain.
 */
const MAX_NAME = 40;

/** The value every entry agrees on, or "" when they do not all agree. */
function agreed(values: readonly (string | undefined)[]): string {
  let found = '';
  for (const value of values) {
    const clean = sanitizeLabel(value ?? '');
    if (clean === '') continue;
    if (found === '') found = clean;
    else if (found !== clean) return '';
  }
  return found;
}

function offersBackend(host: Host, backend: BackendKind): boolean {
  const info = (host.backend_info ?? []).find((entry) => entry.kind === backend);
  return info?.available === true || (host.backends ?? []).includes(backend);
}

/**
 * The parts of a draft a name is made of.
 *
 * Structural rather than the wizard's own PoolDraft, so this module owes the
 * wizard nothing: a draft satisfies it by having the fields, and the naming
 * rules can be read without opening a component.
 */
export interface NameShape {
  backend: BackendKind;
  cpus: string;
  memory_mb: string;
  platform_os: string;
  platform_os_version: string;
  platform_arch: string;
}

/**
 * A version with its dots taken out: 24.04 becomes 2404, 12 stays 12.
 *
 * The same rule as naming.CompactVersion in Go, because the name a pool is
 * suggested here has to be the name `zoomies init` would suggest for the same
 * machine -- an operator who meets both should meet one convention.
 */
function compactVersion(v: string): string {
  return v
    .trim()
    .split(/[._]/)
    .slice(0, 2)
    .map((part) => part.replace(/\D/g, ''))
    .join('');
}

/**
 * What a pool of this draft is made of, most significant part first.
 *
 * This is the shape half of the grammar in docs/naming.md: how much machine,
 * running what. The draft's own fields win, and where it has not said, the
 * connected hosts answer -- but only when every host that could run this pool
 * agrees, because a name that picked one of two answers is a name that lies
 * about half the fleet.
 *
 * Architecture follows the same rule the rest of the grammar does: amd64 is the
 * overwhelming default and saying it on every pool stops the word carrying
 * information, so only the others are spelled out.
 */
export function shape(draft: NameShape, hosts: readonly Host[]): string[] {
  const parts: string[] = [];

  const cpus = Math.ceil(Number(draft.cpus));
  if (Number.isFinite(cpus) && cpus > 0) parts.push(`${cpus}vcpu`);
  const gb = Math.floor(Number(draft.memory_mb) / 1024);
  if (Number.isFinite(gb) && gb > 0) parts.push(`${gb}gb`);

  const offering = hosts.filter((host) => offersBackend(host, draft.backend));
  const from = offering.length > 0 ? offering : hosts;

  const os = sanitizeLabel(draft.platform_os) || agreed(from.map((host) => host.platform?.os));
  if (os !== '') {
    parts.push(os);
    const version =
      compactVersion(draft.platform_os_version) ||
      compactVersion(unanimous(from.map((host) => host.platform?.os_version)));
    if (version !== '') parts.push(version);
  }

  const arch =
    sanitizeLabel(draft.platform_arch) || agreed(from.map((host) => host.platform?.arch));
  if (arch !== '' && arch !== 'amd64') parts.push(arch);

  return parts;
}

/** The raw value every entry agrees on, before sanitising, or "". */
function unanimous(values: readonly (string | undefined)[]): string {
  let found = '';
  for (const value of values) {
    const clean = (value ?? '').trim();
    if (clean === '') continue;
    if (found === '') found = clean;
    else if (found !== clean) return '';
  }
  return found;
}

/**
 * The name for a pool of this draft, and the kennel word it falls back on.
 *
 * A pool is named for its shape, which is the same name `zoomies init` prints
 * for a machine of that size and platform -- so the wizard and the installer
 * stop offering an operator two conventions for the same thing.
 *
 * The kennel word earns its place where the shape cannot carry the name on its
 * own. A pool created before anything about it is known has no shape to be
 * named for, and two pools of the same shape are two pools GitHub needs two
 * names for; in both cases a spaniel tells them apart, which
 * `zoomies-4vcpu-ubuntu-2404-2` never would.
 *
 * Over budget, the least significant parts go first -- architecture, then the
 * release, then the operating system -- because the size is what an operator is
 * choosing between and the word is what they recognise the pool by.
 */
export function poolName(
  word: string,
  draft: NameShape,
  hosts: readonly Host[],
  taken: readonly string[] = [],
): string {
  const parts = named(draft, hosts);
  const spelled = new Set(taken.map((name) => sanitizeLabel(name)));
  const bare = fit(parts);
  if (parts.length > 0 && !spelled.has(sanitizeLabel(bare))) return bare;
  return nicknamedPoolName(word, draft, hosts);
}

/**
 * The same name with a spaniel in it whether it needs one or not.
 *
 * This is what the dice ask for. A pool whose shape is already unique is
 * offered a name without a word, so a button that only sometimes changed
 * anything would read as broken -- and an operator who wants a handle for a
 * pool is entitled to one without having to invent it themselves.
 */
export function nicknamedPoolName(word: string, draft: NameShape, hosts: readonly Host[]): string {
  return fit(named(draft, hosts), sanitizeLabel(word) || 'pool');
}

/**
 * The shape, plus the marker a process-backend pool carries.
 *
 * That backend is a different thing to ask for at the same shape -- its jobs get
 * the host itself rather than a container -- and a workflow author choosing
 * between two pools needs the name to say which. `zoomies init` marks it the
 * same way.
 */
function named(draft: NameShape, hosts: readonly Host[]): string[] {
  const parts = shape(draft, hosts);
  if (draft.backend === 'process') parts.push('host');
  return parts;
}

/**
 * These parts as a branded name, dropping the least significant until it fits.
 *
 * `keep` is never dropped: when it is there at all it is the only thing telling
 * this pool from one of the same shape, so a budget that spent it would produce
 * the name it was added to avoid.
 */
function fit(parts: readonly string[], keep = ''): string {
  const kept = [...parts];
  const tail = keep === '' ? '' : `-${keep}`;
  while (kept.length > 0) {
    const candidate = RUNNER_NAME_PREFIX + kept.join('-') + tail;
    if (candidate.length <= MAX_NAME) return candidate;
    kept.pop();
  }
  return (RUNNER_NAME_PREFIX + keep).slice(0, MAX_NAME).replace(/-+$/, '');
}

/**
 * A word from the kennel, never the one already on screen.
 *
 * Rolling the dice and getting the same name back reads as a broken button,
 * so the previous word is taken out of the hat rather than left to chance.
 */
export function spinWord(previous?: string): string {
  const choices = KENNEL.filter((word) => word !== previous);
  const pool = choices.length > 0 ? choices : KENNEL;
  return pool[Math.floor(Math.random() * pool.length)] ?? 'biscuit';
}

File: internal/naming/names.go
package naming

import (
	"strings"
)

// PoolName is the canonical name for a pool of the given shape.
//
// It carries no suffix: a pool *is* its shape, and two pools with the same
// shape are the same pool. That is also what makes the name usable in a
// workflow's runs-on, which is the name's main job.
func PoolName(spec Spec) string { return spec.WithSuffix("").String() }

// HostName is the canonical name for a machine: its shape, plus enough of the
// machine's own name to tell two identical machines apart.
//
// The machine name goes last because it is the part an operator scans for --
// the eye stops at the end of the line -- and because it is the part most
// likely to be truncated on a host whose hostname is a cloud provider's idea
// of a name.
func HostName(spec Spec, machine string) string {
	return spec.WithSuffix(shortMachine(machine)).String()
}

// Labels are the labels a pool of this shape advertises.
//
// The canonical name comes first: it is the one a workflow author writes, and
// the only one that says how much machine they are asking for. The kernel and
// architecture labels follow because actions/runner advertises them anyway,
// and because a pool that declares them is a pool the scheduler will refuse to
// hand an x64 job to when it is arm64 -- which is the whole point of naming
// the architecture in the first place.
func Labels(spec Spec) []string {
	out := make([]string, 0, 3)
	if name := PoolName(spec); name != Prefix {
		out = append(out, name)
	}
	if k := Kernel(spec.OS); k != "" {
		out = append(out, k)
	}
	if a := ArchLabel(spec.Arch); a != "" {
		out = append(out, a)
	}
	return out
}

// shortMachine reduces a hostname to the part that identifies the machine.
//
// A fully qualified name repeats the same domain on every host in the fleet,
// which is precisely the information a name should not spend characters on.
func shortMachine(machine string) string {
	if host, _, ok := strings.Cut(strings.TrimSpace(machine), "."); ok {
		machine = host
	}
	return Slug(machine)
}

// Base is the branded form of a name that other names are built from: lowercase,
// hyphenated, and carrying the prefix exactly once.
//
// A pool called "gpu" and a pool called "zoomies-gpu" are the same pool as far
// as this grammar is concerned, and neither should produce "zoomies-zoomies-gpu".
func Base(name string) string {
	s := Slug(name)
	switch {
	case s == "":
		return Prefix
	case s == Prefix, strings.HasPrefix(s, Prefix+"-"):
		return s
	}
	return Prefix + "-" + s
}

// RunnerName is the name one runner registers under: what its pool is made of,
// then a discriminator that tells it from its siblings.
//
// GitHub shows this name in the runner list, in the job header, and in the "Set
// up job" step of every job's log -- the three places somebody looks when a job
// went somewhere surprising. "zoomies-4vcpu-ubuntu-2404-biscuit-a3f9qz2m" answers
// "how much machine, running what" there, without a page load; and because the
// shape is rendered by the same grammar as the pool's own name, Parse reads a
// runner's name back into the pool's Spec plus the discriminator.
//
// Two things must survive a name that will not fit. The prefix, because it is
// what marks a registration as this fleet's rather than one somebody made by
// hand -- store.IsRunnerName, and the reaper behind it, know a runner by no
// other sign. And the discriminator, because GitHub requires the name to be
// unique within the target and two runners that truncated to the same string
// would be one registration fighting itself. So the shape in the middle is what
// gives way, a whole segment at a time: a name cut mid-word ("...ubuntu-24")
// claims a platform that does not exist, where a name one segment shorter only
// says less than it could.
func RunnerName(pool, discriminator string) string {
	tail := Slug(discriminator)
	if room := MaxNameLength - len(Prefix) - 1; len(tail) > room {
		tail = strings.TrimRight(tail[:max(room, 0)], "-")
	}
	base := Base(pool)
	if tail == "" {
		return truncate(base)
	}
	parts := strings.Split(base, "-")
	for len(parts) > 1 && len(strings.Join(parts, "-"))+1+len(tail) > MaxNameLength {
		parts = parts[:len(parts)-1]
	}
	return truncate(strings.Join(parts, "-") + "-" + tail)
}
